zeus/monitor/rapl.py

```diff
@@ -0,0 +1,146 @@
+        for more details.
+
+    Attributes:
+    """
+
+    def __init__(
+        self,
+        rapl_file_path: str,
+        max_energy_uj: float,
+        rapl_csv_path: str | None = None,
+    ) -> None:
+        """Initialize the rapl monitor.
+
+        Args:
+            rapl_file_path: File path where the RAPL file is located
+            rapl_csv_path: If given, the wrap around polling will write measurements
+                to this path. Otherwise, a temporary file will be used.
+        """
+        if not os.path.exists(rapl_file_path):
+            raise ValueError(f"{rapl_file_path} is not a valid file path")
+        self.rapl_file_path = rapl_file_path
+
+        # Set up logging.
+        self.logger = get_logger(type(self).__name__)
+
+        self.logger.info(f"Monitoring wrap around of {rapl_file_path}")
+
+        # Create and open the CSV to record power measurements.
+        if rapl_csv_path is None:
+            rapl_csv_path = tempfile.mkstemp(suffix=".csv", text=True)[1]
+        open(rapl_csv_path, "w").close()
+        self.rapl_f = open(rapl_csv_path)
+        self.rapl_df_columns = ["time", "energy"]
+        self.rapl_df = pd.DataFrame(columns=self.rapl_df_columns)
+
+        self.num_wraparounds = 0
+        self.last_time = time()
+
+        # Spawn the power polling process.
+        atexit.register(self._stop)
+        self.process = mp.get_context("spawn").Process(
+            target=_polling_process,
+            args=(rapl_file_path, max_energy_uj, rapl_csv_path),
+        )
+        self.process.start()
+
+    def _stop(self) -> None:
+        """Stop monitoring power usage."""
+        if self.process is not None:
+            self.process.terminate()
+            self.process.join(timeout=1.0)
+            self.process.kill()
+            self.process = None
+
+    def _update_df(self) -> None:
+        """Add rows to the power dataframe from the CSV file."""
+        try:
+            additional_df = typing.cast(
+                pd.DataFrame,
+                pd.read_csv(self.rapl_f, header=None, names=self.rapl_df_columns),
+            )
+        except pd.errors.EmptyDataError:
+            return
+
+        if additional_df.empty:
+            return
+
+        if self.rapl_df.empty:
+            self.rapl_df = additional_df
+        else:
+            self.rapl_df = pd.concat(
+                [self.rapl_df, additional_df],
+                axis=0,
+                ignore_index=True,
+                copy=False,
+            )
+
+    def get_num_wraparounds(self) -> int:
+        self._update_df()
+        curr_time = time()
+
+        # Get entries between last measured time and current time
+        filtered = self.rapl_df[(self.rapl_df['time'] >= self.last_time) & (self.rapl_df['time'] <= curr_time)]
+        for i in range(len(filtered)-1):
+            if filtered.iloc[i+1]['energy'] < filtered.iloc[i]['energy']:
```

Comment:
Iterating through the dataframe in a `for` loop is very inefficient, especially when the monitor has been running for a long time after `self.last_time`. I think it should be possible to use `df.diff()` and count the number of negative values in the resulting series to count the number of wraparounds. Some one-off error fixing may be necessary.

Reply:
Or, you might just detect the wraparounds inside `_polling_process`. Whenever it reads the RAPL counter, it can always cache the previous counter value. So it can directly compare the current and previous values and *only write to the CSV file if it detected a wraparound*. That will significantly simplify `get_num_wraparounds` -- It'll just be `len(df)`. Maybe it doesn't even have to use Pandas or CSV files; a simple integer protected with `mp.Lock` might be sufficient.